Remove in-relation concept from the SMT encoding with arrays

rodrigo7491 opened the pull request:
<!-- Please ensure that your PR includes the following, as needed -->

- [ ] ~Tests added for any new code~
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] ~Documentation added for any new functionality~
- [ ] ~Entry added to [UNRELEASED.md](./UNRELEASED.md) for any new functionality~

This PR removes the "in-relation" concept from the arrays encoding. It adds three new Apalache IR operators that direct the use of SMT `select` and `store` during constraint generation. A factory class, `InOpFactory`, was added to handle the use of the new operators. Uses of `tla.in` that lead to constraint generation were changed to a call to the factory, which returns a `tla.in` expression for the OOPSLA19 encoding and an expression with the new IR operators for the arrays encoding. Three new methods were added to `Z3SolverContext` to handle the new IR operators. See the updated ADR for more details. Closes #1106.

konnov commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/Z3SolverContext.scala:
```diff
@@ -218,45 +195,46 @@ class Z3SolverContext(val config: SolverConfig) extends SolverContext {
+    log(s";; declare edge inclusion $edgeName")
+    log(s"(declare-const $updatedSetName $setSort)")
+    val updatedSet = z3context.mkConst(updatedSetName, setSort)
+    cellCache += (setId -> ((updatedSet, setT, level) :: cellCache(setId)))
```
That's where another cache would make things more obvious :)

rodrigo7491 replied:
As stated in a previous comment, I'm not sure `cellCache` can work without storing a list. If this is the case I'm not sure if it would make sense to have another cache. We could split `cellCache` and have one cache hold the head of the list while the other holds the tail, but I don't see the benefit.

konnov replied:
Yes, I see what you mean now. I am thinking about the non-array encoding. Since it is a list, there should be no big memory penalty for storing the list instead of a single element.

konnov reviewed (commented):
I like this approach. See my comments. We have to be careful about the changes as they are affecting a large set of rewriting rules. That's why I would like to review this PR once again. We should also bump the release version to indicate that potential breaking changes could be introduced. You can do it by running `mvn release:update-versions` and entering `0.18.0-SNAPSHOT`.

konnov commented:
I was also puzzled by the semantics of "unchanged" in many places. Wondering if we could make it more clear.

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/Z3SolverContext.scala:
```diff
@@ -218,45 +195,46 @@ class Z3SolverContext(val config: SolverConfig) extends SolverContext {
+
+  private def mkUnchanged(setId: Int): ExprSort = {
+    val currentset = cellCache(setId).head._1
+    val oldSet = if (cellCache(setId).size > 1) cellCache(setId).tail.head._1 else currentset
+
+    val eqUnchanged = z3context.mkEq(currentset, oldSet)
```
I see that the behaviour of unchanged is not very clear. In the OOPSLA encoding set membership is undefined initially, and it needs to be set to true or false via `\in` or `\notin`. Since all arrays are initialized as the empty set, by default `\A x . x \notin Some_Set` and if membership wants to be asserted a new, updated, set needs to be defined, explicitly adding the element.

That is why (I believe) keeping the array as it is works here. I thought of adding a `not(select(...))` constraint, but I think it would be redundant. Every `\in` and `\notin` has a `read` and a `write` version, as you want to call them, and figuring out which version to use where was a good part of the problems I had with this PR. For `\notin`, its `read` version became `not(mkRead)`, while its `write` version became `mkUnchanged`, as it is already not in the set by default. A problem can arise if we substitute `\notin` by the wrong factory call, but if the rules are all correct, the approach should be sound.

rodrigo7491 replied:
Regarding the trivial case, I noticed it but was not sure if I should simply assert true. Will make the change.

konnov replied:
I think we have to be very careful about "unchanged" when we are removing elements, not adding them. Maybe it is worth adding directly in the comment for unchanged.

rodrigo7491 commented on docs/src/adr/011adr-smt-arrays.md:
```diff
@@ -136,18 +136,24 @@ The following changes will be made to implement the new encoding of sets:
+  - The `storeInSet` IR operator represents the SMT `store`.
+  - The `unchangedSet` IR operator represents an equality between the current and new SSA array
+    representations. This is required to constraint the array representation as it evolves.
+- Add `InOpFactory` to enable the appropriate use of either `TlaSetOper.in` or the new IR operators,
```
I will make the following renaming, as suggested here and in some other comments:
```
InOpFactory -> SetMembershipFactory
mkAccessOp -> mkReadMem
mkUpdateOp -> mkWriteMem
mkUnchangedOp -> mkNotMem
```

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/caches/RecordDomainCache.scala:
```diff
@@ -45,9 +47,9 @@ class RecordDomainCache(solverContext: SolverContext, strValueCache: ModelValueC
         if (usedKeys.contains(key)) {
-          tla.in(cell.toNameEx, set.toNameEx)
+          inOpFactory.mkUpdateOp(cell, set)
         } else {
-          tla.not(tla.in(cell.toNameEx, set.toNameEx))
+          tla.not(inOpFactory.mkAccessOp(cell, set))
```
In the second case we want to make sure that `cell` is not in `set`. Using `mkUnchanged` would not achieve this behaviour.

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rewriter/ConstSimplifierForSmt.scala:
```diff
@@ -52,6 +52,11 @@ class ConstSimplifierForSmt extends ConstSimplifierBase {
       ex
 
+    // same as tla.in
+    case ex @ (OperEx(ApalacheOper.selectInSet, _*) | OperEx(ApalacheOper.storeInSet, _*) |
+        OperEx(ApalacheOper.unchangedSet, _*)) =>
+      ex
```
Isn't this case avoiding the simplification done in the default case below, as is done for the case with `in` and `notin`?

konnov replied:
ah, yes. I missed that!

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/FunCtorRule.scala:
```diff
@@ -68,10 +71,15 @@ class FunCtorRule(rewriter: SymbStateRewriter) extends RewritingRule {
+      val inDomain = inOperFactory.mkAccessOp(domElem, domainCell).typed(BoolT1())
+      val inRelation = inOperFactory.mkUpdateOp(relElem, relation).typed(BoolT1())
+      val expr = if (rewriter.solverContext.config.smtEncoding == "arrays") {
+        val notInRelation = inOperFactory.mkUnchangedOp(relElem, relation).typed(BoolT1())
```
Exactly because they are empty by default. We need `notInRelation` to keep track of the SSA representation of the array. Will add a comment as requested.

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/CardinalityRule.scala:
```diff
@@ -57,7 +59,8 @@ class CardinalityRule(rewriter: SymbStateRewriter) extends RewritingRule {
           // newCounter = counter if hd \notin set \/ \E c \in counted: hd = c /\ c \in set
           arena = arena.appendCell(BoolT())
           val beforePred = arena.topCell
-          val beforeEx = tla.or(tla.notin(hd.toNameEx, set.toNameEx), tla.or(counted.map(eqToOther(hd, _)): _*))
+          val beforeEx =
+            tla.or(inOpFactory.mkUnchangedOp(hd.toNameEx, set.toNameEx), tla.or(counted.map(eqToOther(hd, _)): _*))
```
Will change it to `tla.not(inOpFactory.mkAccessOp(hd, set))`. Thanks for spotting this.

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/RecCtorRule.scala:
```diff
@@ -85,7 +86,8 @@ class RecCtorRule(rewriter: SymbStateRewriter) extends RewritingRule {
         if (extraKeyMap.nonEmpty) {
-          val extraOutsideOfDomain = extraKeyMap.values.map(f => tla.notin(f.toNameEx, domain.toNameEx))
+          val extraOutsideOfDomain =
+            extraKeyMap.values.map(f => tla.not(inOpFactory.mkAccessOp(f.toNameEx, domain.toNameEx)))
```
Here we are simply checking membership, so no need to do any updates.

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetInRule.scala:
```diff
@@ -109,12 +112,12 @@ class SetInRule(rewriter: SymbStateRewriter) extends RewritingRule {
       val inCdm = nextState.asCell
       // BUGFIX: check only those pairs that actually belong to the relation
-      tla.or(tla.notin(pair.toNameEx, relation.toNameEx), tla.and(inDom.toNameEx, inCdm.toNameEx))
+      tla.or(inOpFactory.mkUnchangedOp(pair.toNameEx, relation.toNameEx), tla.and(inDom.toNameEx, inCdm.toNameEx))
```
Will change it to tla.not(inOpFactory.mkAccessOp(pair, relation)). Thanks for spotting this.

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetInclusionRuleWithArrays.scala:
```diff
@@ -45,37 +47,50 @@ class SetInclusionRuleWithArrays(rewriter: SymbStateRewriter) extends SetInclusi
-    val leftElems =
-      if (rightIsPowSet) state.arena.getHas(leftCell).flatMap(state.arena.getHas)
-      else state.arena.getHas(leftCell)
+    val leftElems = state.arena.getHas(leftCell)
+    val leftElemsElems = leftElems.flatMap(state.arena.getHas)
```
There is a distinction, but it is now in the conditional below. I had to rework this method because adding the factory calls exposed a bug. Before, if I remember correctly, we were checking, for `{{1,2},{3,4}} \subseteq SUBSET {1,2,3,4}`, if
```
1 \in {1,2,3,4} /\ 1 \in {{1,2},{3,4}}
2 \in {1,2,3,4} /\ 2 \in {{1,2},{3,4}}
3 \in {1,2,3,4} /\ 3 \in {{1,2},{3,4}}
4 \in {1,2,3,4} /\ 4 \in {{1,2},{3,4}}
```
which is obviously wrong. I though at the time that using `flatMap` would make constraints of the form `1 \in {1,2}`, but that was not happening. Now we recursively check the subset relation, i.e.,
```
{1,2} \subseteq {1,2,3,4}
{3,4} \subseteq {1,2,3,4}
```

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetInclusionRule.scala:
```diff
@@ -60,7 +63,7 @@ class SetInclusionRule(rewriter: SymbStateRewriter) extends RewritingRule {
     def eachElem(state: SymbState, elem: ArenaCell): SymbState = {
       val newState = rewriter.lazyEq.subsetEq(state, elem, powDom)
-      val outOrSubsetEq = tla.or(tla.not(tla.in(elem.toNameEx, leftCell.toNameEx)), newState.ex)
+      val outOrSubsetEq = tla.or(inOpFactory.mkUnchangedOp(elem, leftCell), newState.ex)
```
Will also change this one to not(read(...)). I had an earlier version of `mkUnchangedOp` that worked differently, and it was intended to be used like this, but I reworked it.

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/Z3SolverContext.scala:
```diff
@@ -81,6 +81,7 @@ class Z3SolverContext(val config: SolverConfig) extends SolverContext {
    * Again, the cache carries the context level, to support push and pop.
+   * To support SSA arrays representing sets, the cache stores a list of constants.
    */
   private val cellCache: mutable.Map[Int, List[(ExprSort, CellT, Int)]] =
```
We need not only the immediate predecessor, but all of them. Imagine the case below:
```
(assert (= set_0 ((as const (Array Int Bool)) false)))
(assert (= set_1 (store set_0 5 true)))
(push)
(assert (= set_2 (store set_1 6 true)))
(push)
(assert (= set_3 (store set_2 7 true)))
(assert (= (select set_3 7) true))
(pop 2)
(assert (= (select set_1 7) false)) ; Without the list we would query set_3 here
```
I'm not sure Apalache would create such types of constraints, but I thought it better to be safe on the general case.

konnov replied:
Oh, I see. I have not thought about the push/pop case. Amazing that you have thought about it! Maybe add a comment explaining why you need a list. That was not obvious to me at all :)

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetInclusionRuleWithArrays.scala:
```diff
@@ -45,37 +47,50 @@ class SetInclusionRuleWithArrays(rewriter: SymbStateRewriter) extends SetInclusi
+    if (rightIsPowSet) {
+      def eachElem(state: SymbState, elem: ArenaCell): SymbState = {
+        val newState = subset(state, elem, rightElemOrDomain, false)
+        val outOrSubsetEq = tla.or(inOpFactory.mkUnchangedOp(elem, leftCell), newState.ex)
```
Will also make it not(read(...)).

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/PowSetCtor.scala:
```diff
@@ -27,8 +29,10 @@ class PowSetCtor(rewriter: SymbStateRewriter) {
-        rewriter.solverContext
-          .assertGroundExpr(tla.equiv(tla.in(e.toNameEx, subsetCell.toNameEx), tla.in(e.toNameEx, set.toNameEx)))
+        val inSubset = inOpFactory.mkUpdateOp(e, subsetCell)
+        val notInSubset = inOpFactory.mkUnchangedOp(e, subsetCell)
```
In the non-array encoding `inSubset` should evaluate to `tla.in(e, subsetCell)` and `notInSubset` should evaluate to `tla.not(tla.in(e, subsetCell))`.

konnov replied:
I see. It's because the factory works differently for different types of encoding. Now it makes sense! Perhaps, add a comment about that.

rodrigo7491 commented:
> I was also puzzled by the semantics of "unchanged" in many places. Wondering if we could make it more clear.

Where would be a good place to add more details? The ADR? The factory?

konnov commented:
> > I was also puzzled by the semantics of "unchanged" in many places. Wondering if we could make it more clear.
> 
> Where would be a good place to add more details? The ADR? The factory?

Both. Depending on the path, the interested reader will hit it either in the ADR or in the code directly.

konnov commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/Z3SolverContext.scala:
```diff
@@ -218,45 +195,46 @@ class Z3SolverContext(val config: SolverConfig) extends SolverContext {
+    eqStore.asInstanceOf[ExprSort]
+  }
+
+  private def mkUnchanged(setId: Int): ExprSort = {
```
```suggestion
  private def mkUnchangedSet(setId: Int): ExprSort = {
```

konnov reviewed (commented):
I think I understand what's going on now. Once you have made the changes, ping me. I will have a look again, so we do not let unexpected bugs in the next release by accident.

shonfeder commented:
>  We have to be careful about the changes as they are affecting a large set of rewriting rules. That's why I would like to review this PR once again. 

Is it not possible to protect and isolate such changes behind the feature flag we introduced? The aim with the feature flag is to explore introducing different functionality in parallel without risking breakage to our current set up. I haven't looked at the code yet, but if there are changes to affect all code, e.g., by optimizing or refactoring core functionality, IMO those are best introduced is a separate PR, not in a PR that is focused only on the smt encoding.

> We should also bump the release version to indicate that potential breaking changes could be introduced. You can do it by running `mvn release:update-versions` and entering `0.18.0-SNAPSHOT`.

Let's please leave this to the release process. We should not bump the version numbers in dev PRs imo: instead, record the nature of the change int he changelog, and we'll increment the version umber when a new release is cut.

shonfeder commented:
Just had a quick look at the general shape of the change here: I wonder if it would not be possible to push the feature flag determination higher up, and have it trigger selection of the operators in a more modular way. The fact that the configuration is here being passed to determine just one operator looks a bit like trouble IMO, as it seems to signal the need to add new, configurable operator constructors for every operator.

Could we, instead, figure out a way of making the whole operator set responsive to a configuration, and moved into the DI framework?

konnov commented:
> > We have to be careful about the changes as they are affecting a large set of rewriting rules. That's why I would like to review this PR once again.
> 
> Is it not possible to protect and isolate such changes behind the feature flag we introduced? The aim with the feature flag is to explore introducing different functionality in parallel without risking breakage to our current set up. I haven't looked at the code yet, but if there are changes to affect all code, e.g., by optimizing or refactoring core functionality, IMO those are best introduced is a separate PR, not in a PR that is focused only on the smt encoding.

This idea required Rodrigo to refactor the existing code too. That's why I want to double check it.

konnov commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/LazyEquality.scala:
```diff
@@ -451,8 +453,8 @@ class LazyEquality(rewriter: SymbStateRewriter)
-          val membershipTest = tla.in(keyCell.toNameEx, leftDom.toNameEx)
-          //      newState = rewriter.rewriteUntilDone(newState.setRex(tla.in(keyCell, leftDom))) // the old way
+          val membershipTest = inOpFactory.mkAccessOp(keyCell, leftDom)
+//          newState = rewriter.rewriteUntilDone(newState.setRex(inOpFactory.mkAccessOper(keyCell, leftDom))) // the old way
```
I agree

rodrigo7491 replied:
Will do.

konnov commented:
> > We should also bump the release version to indicate that potential breaking changes could be introduced. You can do it by running `mvn release:update-versions` and entering `0.18.0-SNAPSHOT`.
> 
> Let's please leave this to the release process. We should not bump the version numbers in dev PRs imo: instead, record the nature of the change int he changelog, and we'll increment the version number when a new release is cut.

That makes sense. How can we queue the version increment for the next release?

rodrigo7491 commented:
As Igor said, I'm not sure these changes could be isolated behind the feature flag.

> Could we, instead, figure out a way of making the whole operator set responsive to a configuration, and moved into the DI framework?

What is the DI framework? I believe we should not have make all operators dependent on the selected encoding, but maybe I'm wrong.

> Let's please leave this to the release process. We should not bump the version numbers in dev PRs imo: instead, record the nature of the change int he changelog, and we'll increment the version number when a new release is cut.

Should I add an entry to `UNRELEASED` then? I don't think this PR is a new feature exactly, nor is it a bug fix.

konnov commented:
We should add it to `UNRELEASED.md`, because it's a relatively large change and it may have unexpected impact. So the users should be able to see, if there were any potentially breaking changes.

shonfeder commented:
> We should add it to UNRELEASED.md, because it's a relatively large change and it may have unexpected impact. So the users should be able to see, if there were any potentially breaking changes.

I can understand the caution here, on the other hand, if we are only changing implementation details and not actually changing any functionality or producing any breaking changes, I don't know how much sense it makes to warn users that some mistake might break things. We have our test suites, and that should warn us of any breaking changes. If they don't then we obviously are lacking sufficient coverage and should fix the holes, but I don't know if it makes sense to announce a breaking change *just in case*, when in fact we don't anticipate any change in behavior or functionality.

shonfeder commented:
> How can we queue the version increment for the next release?

We determine the version increment based on the features being released, as recorded in the change log.

rodrigo7491 commented:
@konnov @shonfeder I addressed all comments of Igor's first review, so feel free to look at the code again. I will hold committing a change to `UNRELEASED.md` until we reach an agreement on it.

konnov commented:
> > How can we queue the version increment for the next release?
> 
> We determine the version increment based on the features being released, as recorded in the change log.

Hmmm. Is the decision on version increment exclusively feature-based? I remember that some projects had stable and unstable releases, e.g., Debian. I wonder if we could warn the users about potentially breaking changes in some release, if we are not sure ourselves.

shonfeder commented:
> What is the DI framework? 

That's the "dependency injection" framework. We are using Guice.

> I believe we should not have make all operators dependent on the selected encoding, but maybe I'm wrong.

Do you think it's likely we'll have to change any other operators, or is `\in` special in this regard?

shonfeder commented:
> Hmmm. Is the decision on version increment exclusively feature-based? I remember that some projects had stable and unstable releases, e.g., Debian. I wonder if we could warn the users about potentially breaking changes in some release, if we are not sure ourselves.

If we follow semantic versioning, all the guidelines are there: https://semver.org/

If you want to change the branching strategy to maintain stable and unstable releases, we can definitely do that if we think it's worth the overhead. But we don't currently have support for it.

shonfeder commented:
We've arranged a sync for tomorrow, but I'll add a few quick notes here on my thoughts and motivations.

Fist, to state motivations/aims:

1. We all want (of course) to keep the code well structured and as easy to reason about as possible.
2. The array encoding is an experimental feature. While we want to enable integration of it into the mainline, so we can benefit from it, and reduce risk of merge conflicts or drifts that might make it more difficult for the feature to land into principle use, we want to insulate the rest of the code from getting lots of traces of this feature, especially while it is still in the experimental stage.

Current concerns:

-  re: aim 1 -- This introduces a complication to the current configuration data flow (will explain shortly) and it introduces two different ways of constructing expressions, which are intermixed, without any clear principle governing when one should use which way. 
- re: aim 2 -- This PR has us reading from the config value in dozens of places, and this means that we are not imposing a clean isolation of the feature, but are instead weaving it throughout many different core parts of the code.

I''ll now elaborate on these concerns.

1. Complicating the configuration data flow and introducing additional ways of constructing expressions:

Just considering the use in  the `LazyEquality` class, the current PR requires the follow flow of data in order to determine who to build expressions:

```
                   SymbStateRewriter
                         |
                         |
                      (config)
                         |
                         ▼
Builder ---------> SetMembershipFactory
   |                            |
   +-----> LazyEquality <-------+
```

That is, we use the configured value to configure the `SetMembershipFactory`, but then we also need to draw from `Builder` in the `SetMembershipFactory`, and we draw from `Builder` directly (via the `tla` module and it's `convenience` package) in order to build some expressions, but draw from `SetMembershipFactory` to build other expressions. 

2. Scattering dependency on reading from the config value:

This PR introduces nearly 30 different locations where we are reading from `rewriter.solverContext.config.smtEncoding` accross more than 20 files.  IMO, this is much more infection use of the feature flag than we'd like. If possible, I think we should find a way to consolidate the logic around this.

Here's a rough sketch for a possible alternative approach: 

- Introduce the new operators into the IR as planned (but please see my question below).
- Add the `apalcheSelectInSet`, `apalacheUnchangedSet`, and `apalacheStoreInSet` construcrs to the current `Builder` class, as planed, but just set them as aliases of the current `in` operator.
- Do away with the `SetMembershipFactory`, and instead just call the relevant construct in all relevant places (this will be a no-op change at first).
- (Probably in a followup PR) Add a new `Builder` class, perhaps named `BuilderArrayEncoding`, that extends `Builder` but overrides the new constructors so that they build the newly introduced operators.
- Move selection of the appropriate builder class into the dependency injection pipeline, and wire things up so that the appropriate kind of `Buidler` is selected for all down-stream classes, based on a check of the configured value at one point in the code (ideally near the edge of the pipeline).

Does something like this seam feasible? Does it inspire any related ideas that could help us simplify and better isolate these kinds of changes?

---

One additional question: is introducing these new operators in the IR consistent with the aim of the IR as a distillation of the core of TLA? I worry that it adds complications which are only relevant for the SMT encoding, which, in one sense, I have thought was an implementation detail. Would it not be possible, instead, to infer how to encode the `in` operator into the appropriate SMT constructs, rather than having to explicitly bake this into the encoding?

Kukovec reviewed (commented):
Looks fine, but I'll let Igor give the final approve so it doesn't get closed accidentally or by the automerge.

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetInRule.scala:
```diff
@@ -13,12 +13,13 @@ import at.forsyte.apalache.tla.lir.{NameEx, OperEx, TlaEx}
   override def isApplicable(state: SymbState): Boolean = {
     state.ex match {
-      case OperEx(TlaSetOper.in, NameEx(name), _) =>
+      case OperEx(op, NameEx(name), _) if op == TlaSetOper.in || op == ApalacheOper.selectInSet =>
```
Thanks for the suggestion, I didn't know we could use disjunction like this. I tried making a disjunction between two `OperEx` here, but they could not have the same arg names.

rodrigo7491 commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/PreproSolverContext.scala:
```diff
@@ -124,7 +125,19 @@ class PreproSolverContext(context: SolverContext) extends SolverContext {
         cache.get((left, right)) match {
           case Some(cached) => cached.asTlaEx(negate = false)
-          case None         => ex
+          case None         => tla.in(NameEx(left), NameEx(right))
+        }
+
+      case OperEx(ApalacheOper.selectInSet, NameEx(left), NameEx(right)) =>
+        cache.get((left, right)) match {
+          case Some(cached) => cached.asTlaEx(negate = false)
+          case None         => tla.apalacheSelectInSet(NameEx(left), NameEx(right))
+        }
+
+      case OperEx(ApalacheOper.storeInSet, NameEx(left), NameEx(right)) =>
+        cache.get((left, right)) match {
+          case Some(cached) => cached.asTlaEx(negate = false)
+          case None         => tla.apalacheStoreInSet(NameEx(left), NameEx(right))
```
Will change to `None => ex`, thanks!

konnov commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/SymbStateRewriterImplWithArrays.scala:
```diff
@@ -9,6 +9,16 @@ import at.forsyte.apalache.tla.lir.convenience.tla
 
+/**
+ * This class extends SymbStateRewriterImpl with encoding rules for the encoding with SMT arrays.
+ * TODO: enforce that _solverContext.config.smtEncoding is correct?
```
Just add `require(_solverContext.config.smtEncoding == ...)`?

rodrigo7491 replied:
I initially though of enforcing this at every point the rewriter is constructed, but that is not nice. Adding `require(...)` to the base rewriter and `override require(...)` to the rewriter with arrays does not work, nor does adding `require(cond)` and overriding `cond`. Is there a clean solution to this?

konnov replied:
I thought `require` was just a part of the constructor. Not sure that you can override it.

rodrigo7491 replied:
Exactly. If we want to put `require(cond)` in the parent constructor, we need to make sure `cond` is overwritten before it is accessed. I did a search and it seems that there is no way to do any computation in the child class before the parent constructor is called, so we end up requiring `_solverContext.config.smtEncoding == "oopsla19"` for the arrays encoding.

I could either put the `require` only on the child rewriter or precede the construction of every rewriter with the `require`. Any thoughts?

rodrigo7491 replied:
Will be done in a follow-up PR.

konnov commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/smt/Z3SolverContext.scala:
```diff
@@ -638,6 +605,37 @@ class Z3SolverContext(val config: SolverConfig) extends SolverContext {
         val elemId = ArenaCell.idFromName(elemName)
```
fail when `arraysEnabled` is true?

rodrigo7491 replied:
Yes, but in the arrays encoding `TlaSetOper.in` should never be passed to `Z3SolverContext`.

konnov replied:
yes, that's why we should violate an assertion or throw an exception in this case

rodrigo7491 replied:
With the arrays encoding, as soon as `getInPred` is called an exception is thrown, since `inCache` will be empty. Due to this I though there was no need to throw an exception here.

konnov commented on tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetFilterRule.scala:
```diff
@@ -58,11 +60,13 @@ class SetFilterRule(rewriter: SymbStateRewriter) extends RewritingRule {
-          rewriter.solverContext.assertGroundExpr(ifAndOnlyIf)
+          val inNewSet = tla.apalacheStoreInSet(cell.toNameEx, newSetCell.toNameEx)
+          val notInNewSet =
+            tla.apalacheUnchangedSet(cell.toNameEx, newSetCell.toNameEx) // This ensures that cell is not in newSetCell
```
I still find the name `apalacheUnchangedSet` confusing. It enforces the semantics of the array encoding. In the case of oopsla19 encoding, it is `notin(e, S)`, which is not about the set being unchanged. Shall we call it `apalacheStoreNotInSet`? This would emphasize that this enforces the element to be outside of the set. It just happens that you enforce this constraint as equality in SMT by knowing that a new set is empty.

rodrigo7491 replied:
Sure, I was also not happy with the current name, but could not think of a good alternative. Thanks for the suggestion!

konnov reviewed (commented):
I think we are almost good to merge it in unstable. Just a few comments.

konnov reviewed (approved):
Amazing! Let's merge
